src/sage/modular/drinfeld_modform/ring.py

```diff
@@ -0,0 +1,726 @@
+        else:
+            raise TypeError("names must be None, a comma seperated string "
+                            "or a list of string")
+        return cls.__classcall__(cls, base_ring, rank, group, has_type,
+                                 tuple(names))
+
+    def __init__(self, base_ring, rank, group, has_type, names):
+        self._has_type = has_type
+        self._rank = rank
+        self._base_ring = base_ring
+        q = base_ring.base_ring().cardinality()
+        if has_type:
+            degs = [q**i - 1 for i in range(1, rank, 1)]
+            degs.append((q**rank - 1) / (q - 1))
+        else:
+            degs = [q**i - 1 for i in range(1, rank + 1, 1)]
+        self._poly_ring = PolynomialRing(base_ring, rank, names=names,
+                                         order=TermOrder('wdeglex', degs))
+        self._assign_names(names)
+
+        super().__init__(base=base_ring, category=GradedAlgebras(base_ring))
+
+    def _an_element_(self):
+        r"""
+        Return an element of self.
+
+        TESTS::
+
+            sage: q = 3
+            sage: A = GF(q)['T']
+            sage: K.<T> = Frac(A)
+            sage: M = DrinfeldModularForms(K, 2)
+            sage: M.an_element()
+            g1
+        """
+        return self.element_class(self, self._poly_ring.an_element())
+
+    def _repr_(self):
+        r"""
+        Return the string representation of self.
+
+        TESTS::
+
+            sage: A = GF(2)['T']
+            sage: K = Frac(A)
+            sage: M = DrinfeldModularForms(K, 3)
+            sage: M._repr_()
+            'Ring of Drinfeld modular forms of rank 3 over Fraction Field of Univariate Polynomial Ring in T over Finite Field of size 2 (using GF2X)'
+        """
+        return ("Ring of Drinfeld modular forms of rank %s over %s"
+                % (self._rank, self._base_ring))
+
+    def _generator_coefficient_form(self, i):
+        r"""
+        Return the i-th coefficient form at `T`.
+
+        For internal use only, the user should use
+        :meth:`coefficient_form` instead.
+
+        TESTS::
+
+            sage: q = 3
+            sage: A = GF(q)['T']
+            sage: K.<T> = Frac(A)
+            sage: M = DrinfeldModularForms(K, 3)
+            sage: M._generator_coefficient_form(1)
+            g1
+            sage: M._generator_coefficient_form(2)
+            g2
+            sage: M._generator_coefficient_form(3)
+            g3
+
+        ::
+
+            sage: M = DrinfeldModularForms(K, 3, has_type=True)
+            sage: M._generator_coefficient_form(1)
+            g1
+            sage: M._generator_coefficient_form(2)
+            g2
+            sage: M._generator_coefficient_form(3)
+            h3^2
+        """
+        if self._has_type and i == self.rank():
+            q = self._base_ring.base_ring().cardinality()
+            return self.gen(i-1)**(q - 1)
+        return self.gen(i - 1)
```

Comment:
This shift is a little bit weird. Even more that it is not the same for `gen`.

Reply:
It's simply a matter of indexing and definitions. Usually, the coefficients forms at $T$ are indexed from 1 to $r$ (the $i$-th coefficient form is the coefficient in front of $\tau^i$). 

When I did that a long time ago, I wanted to follow the indexing of multivariate polynomial ring's `gen` method (which starts at zero). Also, I wanted to separate the concepts of "coefficient forms" and "generators" as coefficients form generates the ring, but not all generators are coefficients forms. Maybe it would be nice to add a functionality later were user can specify their own set of generators when instantiating the class similarly to how it is done in `ModularFormsRing`.

However, I do hear you and understand the possible confusion this might arise. What do you think about leaving this like so, but adding a note to the documentation explaining the indexing choice?

Reply:
Okay, I'm fine with this though I think I prefer changing the indexing and starting from 0 (as we do almost all the time in python and sage).